### PR TITLE
Improve TagGroup rendering

### PR DESCRIPTION
**Pull request by reidbarber**
The goal is to reduce flickering in React 18 when a tag group with `maxRows`'s layout changes (i.e. container resize).

An alternate solution involves wrapping the setState in flushSync, but we'll run into issues with calling this from useLayoutEffect.

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Resize the container in stories with `maxRows`. The text content below should not flicker/shift on resize.

### CODE REVIEW

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 1]
```diff
@@ -162,6 +158,17 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
...
+    if (maxRows == null || !isCollapsed || !tagsRef.current) {
+      return undefined;
+    }
+    let tag = tagsRef.current.children[0];
+    let tagStyle = window.getComputedStyle(tag);
+    let maxHeight = (tag.getBoundingClientRect().height + parseInt(tagStyle.marginTop, 10) + parseInt(tagStyle.marginBottom, 10)) * maxRows;
+    return {maxHeight, overflow: 'hidden'};
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isCollapsed, maxRows, tagsRef.current]);
+
```
Comment (reidbarber) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:171]: Bulk of this change is here, and the rest is mostly cleanup.

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 2]
```diff
@@ -107,30 +102,31 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
...
+          }
         }
-        let tagStyle = window.getComputedStyle(tags[0]);
-        let maxHeight = (parseInt(tagStyle.height, 10) + parseInt(tagStyle.marginTop, 10) * 2) * maxRows;
+
         return {
-          visibleTagCount: index,
-          showCollapseButton: index < state.collection.size,
-          maxHeight
+          visibleTagCount: Math.max(index, 1),
```
Comment (reidbarber) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:122]: Added so we never show no tags when collapsed.

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 3]
```diff
@@ -162,6 +158,17 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
   let showActions = tagState.showCollapseButton || (actionLabel && onAction);
   let isEmpty = state.collection.size === 0;
 
+  let containerStyle = useMemo(() => {
+    if (maxRows == null || !isCollapsed || !tagsRef.current) {
+      return undefined;
+    }
+    let tag = tagsRef.current.children[0];
+    let tagStyle = window.getComputedStyle(tag);
+    let maxHeight = (tag.getBoundingClientRect().height + parseInt(tagStyle.marginTop, 10) + parseInt(tagStyle.marginBottom, 10)) * maxRows;
```
Comment (snowystinger) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:167]: this is probably only two values, mobile and desktop, right? can we hard code them?

Comment (reidbarber) [reply packages/@react-spectrum/tag/src/TagGroup.tsx:167]: I suppose if we really wanted to optimize, we could memoize these values and only refresh them if theme changes.

Comment (snowystinger) [reply packages/@react-spectrum/tag/src/TagGroup.tsx:167]: lets do that

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 4]
```diff
@@ -162,6 +170,15 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
...
   let isEmpty = state.collection.size === 0;
 
+  let containerStyle = useMemo(() => {
+    if (maxRows == null || !isCollapsed || !tagsRef.current) {
+      return undefined;
+    }
+    let maxHeight = (TAG_STYLES[scale].height + (TAG_STYLES[scale].margin * 2)) * maxRows;
+    return {maxHeight, overflow: 'hidden'};
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isCollapsed, maxRows, tagsRef.current]);
```
Comment (reidbarber) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:180]: Bulk of this PR's behavior change is here, and the rest is mostly cleanup.

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 5]
```diff
@@ -73,24 +85,19 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
...
         let currTagsRef: HTMLDivElement | null = tagsRef.current;
-        if (!currContainerRef || !currTagsRef) {
-          return;
-        }
-
-        let tags = [...currTagsRef.children];
-        let buttons = [...currContainerRef.parentElement.querySelectorAll('button')];
-        if (tags.length === 0 || buttons.length === 0) {
+        let tags = currTagsRef ? [...currTagsRef.children] : [];
+        if (!currContainerRef || tags.length === 0) {
```
Comment (LFDanLu) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:89]: Can we check `state.collection.size === 0` here instead? The render empty state div is a child of `currTagsRef` so `tags.length` won't ever be 0.

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 6]
```diff
@@ -107,36 +114,37 @@ function TagGroup<T extends object>(props: SpectrumTagGroupProps<T>, ref: DOMRef
...
 
         // Remove tags until there is space for the collapse button and action button (if present) on the last row.
-        let buttonsWidth = buttons.reduce((acc, curr) => acc += curr.getBoundingClientRect().width, 0);
-        buttonsWidth += parseInt(window.getComputedStyle(buttons[buttons.length - 1]).marginRight, 10) * 2;
-        let end = direction === 'ltr' ? 'right' : 'left';
-        let containerEnd = currContainerRef.parentElement.getBoundingClientRect()[end];
-        let lastTagEnd = tags[index - 1]?.getBoundingClientRect()[end];
-        lastTagEnd += parseInt(window.getComputedStyle(tags[index - 1]).marginRight, 10);
-        let availableWidth = containerEnd - lastTagEnd;
+        let buttons = [...currContainerRef.parentElement.querySelectorAll('button')];
```
Comment (LFDanLu) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:117]: Was a bit nervous about this `querySelectorAll` but should be fine if we early return for empty tag groups

**Review by LFDanLu — CHANGES_REQUESTED** [review]
One small thing I noticed code wise but otherwise confirmed that the text below the tags didn't get pushed down when the tags flickering happens

File: packages/@react-spectrum/tag/src/TagGroup.tsx (hunk 4 above)
Comment (snowystinger) [reply packages/@react-spectrum/tag/src/TagGroup.tsx:180]: we should remove the tagsRef.current from the dependency array, only state and derived values should be in a dependency array

why do we need to check it, could we check something else instead?

Comment (reidbarber) [reply packages/@react-spectrum/tag/src/TagGroup.tsx:180]: I think that was just leftover from measuring there, I can remove that. Thanks

File: packages/@react-spectrum/tag/src/TagGroup.tsx [hunk 7]
```diff
@@ -19,14 +19,25 @@ import {FocusRing, FocusScope} from '@react-aria/focus';
...
-import {Provider, useProviderProps} from '@react-spectrum/provider';
+import {Provider, useProvider, useProviderProps} from '@react-spectrum/provider';
 import React, {ReactElement, useCallback, useEffect, useMemo, useRef, useState} from 'react';
 import styles from '@adobe/spectrum-css-temp/components/tags/vars.css';
 import {Tag} from './Tag';
 import {useFormProps} from '@react-spectrum/form';
 import {useId, useLayoutEffect, useResizeObserver, useValueEffect} from '@react-aria/utils';
 import {useLocale, useLocalizedStringFormatter} from '@react-aria/i18n';
 
+const TAG_STYLES = {
```
Comment (ktabors) [inline packages/@react-spectrum/tag/src/TagGroup.tsx:30]: If the css changes this breaks, right? Shouldn't there be a comment on the CSS and this making that connection?

**Review by ktabors — COMMENTED** [review]
Works a lot better. There is one case where I can see a flicker, but when I do a screen recording the flickering isn't captured.

**Review by snowystinger — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]

File: packages/@react-spectrum/tag/src/TagGroup.tsx (hunk 7 above)
Comment (reidbarber) [reply packages/@react-spectrum/tag/src/TagGroup.tsx:30]: True, but I don't think these change too often, and we do the same type of thing in TableView and ListView. I think if we have a big enough change to styles, we'll notice here.

**Pull request merged**